typed-racket-more/typed/rackunit/main.rkt

```diff
@@ -93,22 +94,108 @@
   [t:check-around check-around]
   [t:current-check-handler current-check-handler]
   [t:current-check-around current-check-around])
-(define-rewriter t:test-case test-case
-  [t:current-test-case-around current-test-case-around]
-  [t:check-around check-around]
-  [t:current-check-handler current-check-handler]
-  [t:current-check-around current-check-around])
+
+(define-syntax-rule (test-case name expr ...)
+  (parameterize
+      ([current-test-name (ann name String)])
+    (test-begin expr ...)))
+
 (provide test-begin test-case)
 
 (require/opaque-type TestCase test-case? rackunit)
 (provide TestCase test-case?)
 
+(require/typed
+ rackunit/private/monad
+ [#:opaque monad monad?])
+(define-type Seed (Option monad))
 
+(define-type test-suite-handler-down
+  (rackunit-test-suite (Option String) (Thunk Any) (Thunk Any) Seed -> Seed))
+(define-type test-suite-handler-up
+  (rackunit-test-suite (Option String) (Thunk Any) (Thunk Any) Seed Seed -> Seed))
+(define-type test-suite-handler-here
+  (rackunit-test-case (Option String) (Thunk Any) Seed -> Seed))
 
-(require (only-in rackunit test-suite))
+(require/typed
+ rackunit
+ [#:struct rackunit-test-case ([name : (Option String)] [action : (Thunk Any)])
+           #:constructor-name make-rackunit-test-case]
+ [#:struct rackunit-test-suite
+           ([name : String]
+            [tests : (test-suite-handler-down
+                      test-suite-handler-up
+                      test-suite-handler-here
+                      Seed -> Seed)]
+            [before : (Thunk Any)]
+            [after : (Thunk Any)])
+           #:constructor-name make-rackunit-test-suite])
+(require/typed
+ rackunit/private/test-suite
+ [apply-test-suite (rackunit-test-suite
+                    test-suite-handler-down
+                    test-suite-handler-up
+                    test-suite-handler-here
+                    Seed -> Seed)])
+
+(define current-seed : (Parameter Seed)
+  (make-parameter #f))
+
+(: test-suite-test-case-around (test-suite-handler-here -> ((Thunk Any) -> Any)))
+(define (test-suite-test-case-around fhere)
+  (lambda (thunk)
+    (let* ([name (current-test-name)]
+           [test (make-rackunit-test-case name thunk)]
+           [seed (current-seed)])
+      (current-seed (fhere test name thunk seed)))))
+
+(: test-suite-check-around (test-suite-handler-here -> ((Thunk Any) -> Any)))
+(define (test-suite-check-around fhere)
+  (lambda (thunk)
+    (let* ([name #f]
+           [test (make-rackunit-test-case name thunk)]
+           [seed (current-seed)])
+      (current-seed (fhere test name thunk seed)))))
+
+(define-syntax (test-suite stx)
```

Comment:
Add a comment saying that this is copy-pasted, so we know what to refactor later.